src/framework/preferences/js/SelfVoicingEnactor.js

```diff
@@ -92,45 +108,280 @@ var fluid_3_0_0 = fluid_3_0_0 || {};
-    // This should be expanded to read other text descriptors as well.
-    fluid.prefs.enactor.selfVoicing.readFromDOM = function (that, elm) {
-        elm = $(elm);
-        var nodes = elm.contents();
-        fluid.each(nodes, function (node) {
-            if (node.nodeType === fluid.prefs.enactor.selfVoicing.nodeType.TEXT_NODE && node.nodeValue) {
-                that.tts.queueSpeech(node.nodeValue);
-            }
+    /**
+     * Tests if a string is a word, that is it has a value and is not only whitespace.
+     * inspired by https://stackoverflow.com/a/2031143
+     *
+     * @param {String} str - the String to test
+     *
+     * @return {Boolean} - `true` if a word, `false` otherwise.
+     */
+    fluid.prefs.enactor.selfVoicing.isWord = function (str) {
+        return fluid.isValue(str) && /\S/.test(str);
+    };
 
-            if (node.nodeType === fluid.prefs.enactor.selfVoicing.nodeType.ELEMENT_NODE && window.getComputedStyle(node).display !== "none") {
-                if (node.nodeName === "IMG") {
-                    var altText = node.getAttribute("alt");
-                    if (altText) {
-                        that.tts.queueSpeech(altText);
+    /**
+     * Determines if there is rendered text to in an element.
+     * Will return false in the following conditions:
+     * - elm is falsey (undefined, null, etc.)
+     * - elm's offsetHeight is 0 (e.g. display none set on itself or its parent)
+     * - elm has no text or only whitespace
+     * - elm or its parent has `aria-hidden="true"` set.
+     *
+     * @param {jQuery|element} elm - either a DOM node or a jQuery element
+     *
+     * @return {Boolean} - returns true if there is rendered text within the element and false otherwise. (See rules above)
+     */
+    fluid.prefs.enactor.selfVoicing.hasRenderedText = function (elm) {
+        elm = fluid.unwrap(elm);
+
+        return elm && !!elm.offsetHeight && fluid.prefs.enactor.selfVoicing.isWord(elm.innerText) && !$(elm).closest("[aria-hidden=\"true\"]").length;
+    };
+
+    /**
+     * Recursively parses a DOM element and it's sub elements to construct an array of data points representing the
+     * words and space between the words. This data structure provides the means for locating text to highlight as the
+     * self voicing engine runs.
+     * NOTE: consecutive whitespace is collapsed to the first whitespace character.
+     * NOTE: hidden text is skipped.
+     *
+     * @param {jQuery|element} elm - the DOM node to parse
+     * @param {Number} blockIndex - The `blockIndex` represents the index into the entire block of text being parsed.
+     *                              It defaults to 0 and is primarily used internally for recursive calls.
+     *
+     * @return {Array} - An array of data points, objects of the with the following structure.
+     *                   {
+                             blockIndex: {Number}, // the index into the entire block of text being parsed
```

Comment:
We tend to be more specific when we can with "Integer" when we know that the value can't be an arbitrary number